### DISCUSSION TITLE
Question: Where is the difference between: "Microsoft.Toolkit.HighPerformance.Buffers.SpanOwner<T>" and "DotNext.MemoryRental<T>"?

### DISCUSSION THREAD

**Discussion by CodingMadness** [dotnet/dotNext#159] [Q&A] [answered]
Title.

src for "SpanOwner<T>": https://learn.microsoft.com/en-us/dotnet/communitytoolkit/high-performance/spanowner

I cant quite grasp in what scenario yours may be better or equal to that type? I liked from your type that it may rent from whatever pool depending on the size needed, although I dont know if the reasoning-algorithm for that beats the purpose of "SpanOwner<T>" from "Microsoft.Toolkit.HighPerformance" Nuget-package. 

Any advise to this is very very appreciated cause I am in process to determine the proper buffering-strategy and would be cool to get the best one selected ^^

Btw very strong and dedicated work you have put in here, you earned my highest respect!

**Comment by sakno** [comment]
`SpanOwner<T>` is a **stack-only buffer type** according to its description, i.e. **ref** type. It cannot be used inside of **async** methods. The closest counterpart from .NEXT is [MemoryRental&lt;T&gt;](https://dotnet.github.io/dotNext/api/DotNext.Buffers.MemoryRental-1.html) **ref** type. But, at the closer look, they are different. `MemoryRental<T>` simplifies an implementation of memory allocation strategy when you need to choose between stack-allocated memory (for small buffers) and large buffers:

```csharp
const int stackallocThreshold = 20;
var memory = size <=stackallocThreshold
  ? new MemoryRental<byte>(stackalloc byte[stackallocThreshold], size)
  : new MemoryRental<byte>(size);
```

You can find a lot of places in .NEXT codebase that use this pattern.

[MemoryOwner&lt;T&gt;](https://dotnet.github.io/dotNext/api/DotNext.Buffers.MemoryOwner-1.html) is a regular value type (not **ref**) that can be used inside of **async** context. Thus, it can't be used to make abstraction over stack-allocated memory. Its counterpart from Community Toolkit is [MemoryOwner&lt;T&gt;](https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.highperformance.buffers.memoryowner-1), surprise :smile: (I never tried to copy-paste from Community Toolkit implementation, moreover, I even didn't know about it in the beginning of .NEXT project). However, implementation from .NEXT is different:
- It supports `IMemoryOwner<T>` interface as a rented memory holder in addition to custom `ArrayPool<T>`
- As a result, it allows to use any memory that can be represented by `T[]`, `ArrayPool<T>` or `IMemoryOwner<T>`
- Typically, it is combined with [MemoryAllocator&lt;T&gt;](https://dotnet.github.io/dotNext/api/DotNext.Buffers.MemoryAllocator-1.html) delegate type that represents extension point to the caller code and allows to provide any custom memory rental strategy. For instance, `DotNext.Unsafe` library exposes unmanaged memory allocator that allocates the memory out of GC control. And it still uses `MemoryOwner<T>` to represent the allocated memory.
- It supports indexer
- It supports realloc

**Reply by sakno** [reply]
Both types provide low-level API to work with rented memory. If you are interested in growable (dynamic) buffers built on top of these types I recommend you to take a look at [this article](https://dotnet.github.io/dotNext/features/io/buffers.html).

**Comment by CodingMadness** [comment]
First, thx alot for the detailed explanation!

Would there be 1) any benefit, in actually creating a hybrid buffer, who takes any give (sizeOf(T) * count) and checks of HOW MUCH of that (sizeOf(T) * count) does actually fit inside a reasonable stackalloc (lets say 1-2 KB) and THE REMAINING of that size-expression will be rented from a shared pool?

example:
```

const int stackallocThreshold = 2048;
int sizeOfTInBytes= (sizeOf(T) * count);

int diff = Math.Abs(stackallocThreshold  - sizeOfTInBytes);

   //hypothetical ctor()...!
  ///AllocationOptions= AllocationOptions.StackOnly,   AllocationOptions.PoolOnly,   AllocationOptions.UseBoth
  ///it will link both buffers internally 'somehow'
  //'diff' shall be rented from the pool and 'sizeOfTInBytes' on the stack via stackalloc
  //both memory-segments shall be linked internally to work for the user seemingly as if it would be a contigeous buffer in either stack or pool
using MemoryRental<T> result = sizeOfTInBytes<= stackallocThreshold ? 
     new MemoryRental<T>(stackalloc T[stackallocThreshold], count) : 
     new MemoryRental<T>(AllocationOptions.UseBoth, diff,  sizeOfTInBytes);
```

Would this approach yield any benefit in your opinion? 

Best regards!

**Reply by sakno** [reply]
No, it's not practical. If stack-allocated memory is not enough, just rent a new memory and copy from the stack. Keeping the whole buffer in stack and rented memory will prevent you from efficient implementation of `CopyTo(Span<T>)` method. In that case, you need to copy the elements from the stack and then copy the rest of the elements from the rented memory. If initial capacity doesn't meet your expectations, just increase it to avoid memory rental. However, if overflow happens, the final buffer typically is significantly larger that the initial capacity. In that case, no need to keep the head of the buffer in the separated (stack-allocated) memory.

You can take a look at [BufferWriterSlim&lt;T&gt;](https://dotnet.github.io/dotNext/api/DotNext.Buffers.BufferWriterSlim-1.html) implementation. It can use stack-allocated initial buffer, but if it overflows, the buffer switches to the pool.

**Reply by sakno** [reply]
Also, `const int stackallocThreshold = 2048;` is bad for generic `T` because you don't know `sizeof(T)`. `2048` is a huge value which may cause `StackOverflowException`. The same for `sizeof(T) * count`. `OverflowException`, in case of `sizeof(int) * int.MaxValue`

**Reply by sakno** [reply]
Threshold must be computed dynamically to avoid stack overflow (however, it is turned into a constant by JIT):

https://github.com/dotnet/dotNext/blob/b0e32ad95ec1352bc1565af2ade4a53a75745ef7/src/DotNext/Buffers/MemoryRental.cs#L37-L40

**Reply by CodingMadness** [reply]
Hmm, why is 2KB stackalloc alot, I mean we have 4MB stacksize I believe? If u dont do nasty recursion or your function is not like 2 miles long (very large functions are usually denoted as bad practice anyway) you should never end up having  "Stackoverflow".

**Reply by CodingMadness** [reply]
and btw, I do ofc "2048 / sizeOf(T) " which gives you the accepted byteCount,

**Reply by sakno** [reply]
Max stack size depends on OS. On Windows, it is 1 MB.

OS | Process Stack Size | Thread Stack Size
-- | -- | --
Darwin (macOS, iOS, etc) | 8 MiB | 512 KiB
FreeBSD | 8 MiB | 2 MiB
OpenBSD (before 4.6) | 8 MiB | 64 KiB
OpenBSD (4.6 and later) | 8 MiB | 512 KiB
Windows | 1 MiB | 1 MiB
Alpine 3.10 and older | 8 MiB | 80 KiB
Alpine 3.11 and newer | 8 MiB | 128 KiB
GNU/Linux | 8 MiB | 8 MiB

And you don't know how your code will be used by potential consumer, e.g. you don't know how many frames and stack-allocated space on top of the stack already.

**Reply by CodingMadness** [reply]
Okay wow, I dont know to this day why only windows decide for such bad stack utilization... Nevermind

I have one last question, where is the difference between your MemoryRental<T> and BufferWriterSlim<T> they seem to do pretty much the same, no? I have the feeling with some other buffers too

**Reply by sakno** [reply, marked as answer]
`BufferWriterSlim` is a growable dynamic buffer while `MemoryRental` not. You can add as many elements as you want to `BufferWriterSlim`. It transparently reallocates the internal buffer. `MemoryRental` can't grow. They have the same difference as `List<T>` and `T[]`.

**Reply by CodingMadness** [reply]
yea well I fully oversaw the growing aspect of it lol, my appologize, okay i have all infos, thx alot dude! keep going with this lib you are doing really well with it and I cant thank you enough for it !

**Reply by sakno** [reply]
>  cant thank you enough for it 

There is a way to do that :smile: Just mark this discussion as answered
